### PR TITLE
Elegant pool drain fix

### PR DESCRIPTION
**Pull request by geraldog**
This fixes the Pool draining fast issue I've been having with aiosonic in a more elegant manner than my previous PRs.

The first commit fixes edge cases where Connector doesn't put Connection back into pool during errors.

The second commit converts CyclicQueuePool release method to blocking (i.e. non-async) since it only does that one put_nowait() call and that itself is non-blocking.

Finally, the third commit gets rid of create_task with no guarantee of execution in HttpResponse __del__ and instead calls the Pool release directly in a blocking context thanks to the magic of the second commit.

Third commit also adds Connection to Response regardless of keep-alive set. While this may seem overblown, I'm pretty sure I saw some edge cases of Connection Blocked without Keep-Alive set (misconfigured Servers?)

### CODE REVIEW

**Comment by geraldog** [comment]
@sonic182 even though all checks have passed I have a problem with my crawler with the proposed patches. It raises QueueFull on program exit.

**Comment by sonic182** [comment]
For now this is a first fix (the dns stuff and conn releases)

I think this can worth, I'm gonna check it later after work @geraldog

**Comment by geraldog** [comment]
Thank you @sonic182 and I await anxiously for your review.

Regarding the QueueFull error I was having on program exit, I think I traced the problem to another missing *await* on CyclicQueuePool class, cleanup() method.

I proposed instead that we don't use *await* there but get_nowait(). In my case I was mistakenly calling both client.shutdown() and connector.cleanup() and adding the *await* blocked program exit forever.

get_nowait() did the most correct thing, which was raising QueueEmpty on over-cleanup. Of course, now that I corrected my code I don't do double cleanup anymore, and the program exits correctly.

I'll push a fix commit in an instant.

File: aiosonic/__init__.py [hunk 1]
```diff
@@ -474,6 +474,7 @@ async def _do_request(
             response._set_connection(connection)
         else:
             connection.keep = False
+            response._set_connection(connection)
```
Comment (sonic182) [inline aiosonic/__init__.py:477]: I think this can be just outside the if/else because it is always called

File: aiosonic/__init__.py [hunk 2]
```diff
@@ -222,11 +222,11 @@ async def read_chunks(self) -> AsyncIterator[bytes]:
...
-        if self.chunked and not self.chunks_readed:
-            loop = None
-            if self.connection:
-                loop = get_loop()
-                loop.create_task(self.connection.release())
+        if self.connection and self.connection.blocked:
+            if self.connection.writer:
+               self.connection.writer._transport.abort()
+            self.connection.blocked = False
+            self.connection.connector.pool.release(self.connection)
```
Comment (sonic182) [inline aiosonic/__init__.py:229]: I don't like to call the pool.release directly but for now it is ok, gonna try to refactor it

File: aiosonic/__init__.py [hunk 3]
```diff
@@ -222,11 +222,11 @@ async def read_chunks(self) -> AsyncIterator[bytes]:
...
     def __del__(self):
         # clean it
-        if self.chunked and not self.chunks_readed:
-            loop = None
-            if self.connection:
-                loop = get_loop()
-                loop.create_task(self.connection.release())
+        if self.connection and self.connection.blocked:
+            if self.connection.writer:
+               self.connection.writer._transport.abort()
```
Comment (sonic182) [inline aiosonic/__init__.py:227]: these two stuffs can be inside connection class

**Review by sonic182 — APPROVED** [review]
some comments but still I'm gonna merge and try to improve some stuffs 👍

File: aiosonic/__init__.py (hunk 2 above)
Comment (geraldog) [reply aiosonic/__init__.py:229]: Both CyclicQueuePool and SmartPool now have non-async release methods. We could convert the whole chain of release() inside Connection class and Connector class to non-async perhaps? So we can call release on the Connection from a blocking context.

File: aiosonic/__init__.py (hunk 3 above)
Comment (geraldog) [reply aiosonic/__init__.py:227]: Oh, I see. Good catch indeed

**Comment by geraldog** [comment]
Thanks @sonic182 !

File: aiosonic/__init__.py (hunk 2 above)
Comment (sonic182) [reply aiosonic/__init__.py:229]: Yes, I'm working on a PR just now https://github.com/sonic182/aiosonic/pull/457

**Pull request merged**